### ISSUE TITLE
Thử nghiệm những tham số mới

### ISSUE THREAD

**Issue by Khuat-Duc-Minh** [yenq89/SE2025-14.2#34]
# Bối cảnh
Quá trình huấn luyện (như #31 #32) yêu cầu việc thử nhiều bộ tham số để mô hình chạy, huấn luyện dựa theo các bộ dữ liệu ảnh-văn bản được chọn lọc, xác định. Quá trình này gặp nhiều khó khăn khi kết quả của nhiều lần thử nghiệm chưa thực sự như ý. Xét theo năng lực của mô hình, vẫn còn rất nhiều cách để điều chỉnh tham số, mở ra cơ hội huấn luyện những mô hình hiệu quả và phù hợp với thực tế hơn. Trong issue này, tôi sẽ trình bày hai nội dung chính:
- Toàn bộ tất cả tham số được công khai của mô hình Stable Diffusion khi train LoRA.
- Trình bày thêm một số thông tin về tham số ***Rank*** trong huấn luyện LoRA.

# Tất cả tham số của Stable Diffusion khi train LoRA:
Tất cả tham số và nguồn gốc tìm ra sẽ được trình bày trong hai file `.pdf` và `.docx`. Hai file lần lượt cung cấp mô tả ngắn gọn chức năng của từng tham số, và bản gốc giới thiệu về từng tham số và công dụng. Tôi sẽ bổ sung giá trị mặc định của từng tham số nếu có thể

[StableDiffusion_LoRA_Parameters.pdf](https://github.com/user-attachments/files/24272207/StableDiffusion_LoRA_Parameters.pdf)
[SD_LoRA_Parameters.docx](https://github.com/user-attachments/files/24272208/SD_LoRA_Parameters.docx)

# Tham số ***Rank*** khi train LoRA:
## Train LoRA:
LoRA là một kỹ thuật fine-tune. Đây là một phương pháp này hiệu quả đối với các mô hình Deep Learning, đặc biệt là các mô hình ngôn ngữ lớn (nguồn: https://aivietnam.edu.vn/blog/finetune-lora-llms#gi%E1%BB%9Bi-thi%E1%BB%87u). Hiểu đơn giản cách thức hoạt động như sau, thay vì ở mỗi lần cập nhật, thông tin cập nhật sẽ được chia tách, thu nhỏ thay vì giữ nguyên để giảm bộ nhớ. Điều này thực sự hữu dụng khi các mô hình lớn yêu cầu khối lượng bộ nhớ lên đến hàng TB trong khi LoRA yêu cầu bộ nhớ nhỏ hơn đáng kể mà vẫn giữ được chất lượng của kết quả so với phương pháp ban đầu. Nhìn hình ảnh, khi cần cập nhật, các thông tin (ma trận) nhỏ sẽ được tổng hợp (nhân ma trận) để tạo thành thông tin cập nhật có kích cỡ tương ứng kích cỡ của thông tin trọng số mô hình.

<img width="825" height="347" alt="Image" src="https://github.com/user-attachments/assets/e2ee6296-e753-4f18-807e-4761b7438502" />

Nguồn ảnh (thứ cấp) (mạng): https://viblo.asia/p/llm-101-thao-luan-ve-finetune-model-llm-su-dung-lora-low-rank-adaptation-Yym40YQWJ91

## Rank:
Theo chính tác giả của phương pháp LoRA - Edward Hu, chất lượng của quá trình huấn luyện là đảm bảo và quan trọng hơn một trong những ưu điểm của LoRA là nhà nghiên cứu có thể xác định hướng đi tiếp theo. Cụ thể, khi mô hình *hoạt động kém* (underperform), nhà nghiên cứu cần điều chỉnh thêm *tham số* và tăng *hạng*. Nguyên gốc câu thoại như sau:
> "... I want to point out that one advantage of  LoRA is that it's clear what to do next. If it underperforms, we adapt more  parameters and increase the rank..."

Nguồn: https://www.youtube.com/watch?v=DhRoTONcyZE

Theo một nguồn khác, từ youtuber Mark Hennings (https://www.youtube.com/watch?v=t1caDsMzWBk), việc tăng ***Rank*** giúp tăng chính xác (Precision) (4p40). Nhưng cũng theo ông, để chọn một giá trị phù hợp, thì về lý thuyết, điều đó thực sự không đáng kể để bàn (6p42). Tất nhiên, theo ông, giá trị ***Rank*** cao phù hợp cho việc dạy các hành vi phức tạp, các hành vi trái ngược hoặc nằm ngoại phạm vi train ban đầu (7p23).

Ngoài ra, còn thông số ***Alpha*** liên quan đến ***Rank*** nhưng tôi chưa tìm hiểu đủ nhiều, tôi có thể sẽ bổ sung sau.

# Kết luận:
Tôi cho rằng, nằm trong phạm vi hiểu biết, mỗi người có thể thử và đề xuất cho cả nhóm để cải thiện hiệu quả huấn luyện. Ngoài ra, tham số ***Rank*** của mô hình có thể có **tiềm năng** giúp mô hình **tốt hơn**.
